### PR TITLE
[TAO-2570][Feature] Add normalized compatible-positive SigLIP2 masking

### PR DESCRIPTION
**Pull request by yangchris11**
### What changes are proposed in this pull request?

This PR adds metadata-compatible positive masking to the SigLIP loss used by SigLIP2 CLIP training.

It introduces two positive mask modes:

- `attribute_match_positive`
- `attribute_plus_accessory_match_positive`

Compatible off-diagonal image-text pairs are promoted from negatives to weighted positives. The aligned image-caption pair remains a normal positive with weight `1.0`.

Two weighting strategies are supported:

- `per_pair`: assigns `compatible_positive_weight` to every promoted pair.
- `per_query`: distributes `compatible_positive_weight` across all compatible images for each text query.

In distributed `gather` mode, text features and metadata are gathered globally while image rows remain local. Per-query compatible-pair counts are reduced across ranks in FP32 to preserve exact counts under BF16 training.

Queries without usable metadata evidence do not promote off-diagonal pairs and retain their aligned positive.

Example:

```yaml
train:
  loss_type: siglip
  siglip_loss_dist_impl: gather
  siglip_loss_mask_mode: attribute_plus_accessory_match_positive
  compatible_positive_weight: 0.5
  compatible_positive_normalization: per_query

dataset:
  train:
    type: custom
    include_attribute_metadata: true
```

Positive mask modes currently require exactly one custom training dataset with attribute metadata enabled.

Existing ignore modes remain supported:

- `attribute_match_ignore`
- `attribute_plus_accessory_match_ignore`

### Why are the changes needed?

Standard SigLIP treats every off-diagonal image-text pair as a negative. In datasets containing semantically compatible captions or duplicated attribute combinations, this can penalize valid matches as false negatives.

Positive masking allows compatible off-diagonal pairs to contribute as positives. Per-query normalization prevents queries with many compatible images from contributing disproportionately more positive weight than queries with fewer matches.

The FP32 compatible-count reduction is required for mixed-precision correctness because BF16 cannot exactly represent every integer above 256.

Experiment context motivating the recommended setting:

| Configuration | Best validation mAP |
|---|---:|
| Ignore masking baseline | 85.1164% |
| Positive `0.5`, `per_query`, PA disabled | 85.7938% |
| Positive `1.0`, `per_query`, PA disabled | 85.7041% |
| Positive `2.0`, `per_query`, PA disabled | 85.4893% |

Based on these experiments, `0.5/per_query` is the recommended experimental setting. PA loss did not materially improve the held-out result and is intentionally outside this PR.

These measurements predate the latest metadata-evidence and BF16 corrections and should not be interpreted as a completed reproduction on the final commit.

### Related issues

TAO-2570

### Does this PR introduce _any_ user-facing change?

Yes.

Users can configure compatible-positive behavior through:

```yaml
train.siglip_loss_mask_mode
train.compatible_positive_weight
train.compatible_positive_normalization
```

The default remains:

```yaml
train:
  siglip_loss_mask_mode: none
  compatible_positive_weight: 1.0
  compatible_positive_normalization: per_query
```

Because the default mask mode is `none`, existing configurations retain their previous loss behavior and require no migration. Positive masking defaults to `per_query`; set `compatible_positive_normalization: per_pair` explicitly to apply the full configured weight to every promoted pair.

Users opting into a metadata mask mode must enable:

```yaml
dataset:
  train:
    include_attribute_metadata: true
```

Unsupported dataset and distributed configurations produce validation errors.

### How was this patch tested?

Targeted BF16 normalization and distributed positive-mode tests:

```bash
pytest -q tests/multimodal_unit_test/clip/test_masked_siglip_loss.py \
  -k "per_query_count_uses_float32 or gather_per_query_normalizes_with_global_compatible_counts or gather_positive_per_query_matches_global_reference"
```

Result:

```text
3 passed, 24 deselected
```

The targeted tests verify:

- Compatible-pair counts are accumulated in FP32 when labels use BF16.
- A count of 257 is not rounded to 256 before normalization.
- Distributed `per_query` normalization uses global compatible-pair counts.
- A real two-rank Gloo process group executes text and metadata gathers and the compatible-count reduction.
- Distributed targets, valid masks, weights, mean loss, and image/text gradients match a single-process global reference.
- Compatible pairs are placed across rank boundaries with two promoted pairs per query, producing a weight of `0.5 / 2 = 0.25` per pair.

The complete masked SigLIP loss test file was also run:

```bash
pytest -q tests/multimodal_unit_test/clip/test_masked_siglip_loss.py
```

Result:

```text
29 passed in 6.74s
```

This includes coverage for:

- Normal diagonal positives
- Positive and ignore mask modes
- Attribute and accessory compatibility
- `per_pair` and `per_query` normalization
- Queries without metadata evidence
- Queries with no compatible pairs
- Cross-rank target shapes and positive offsets
- Cross-rank text-gradient propagation
- Unequal distributed batch-size rejection
- Real two-rank positive-mode loss and gradient equivalence
- Single compatibility-mask construction in ignore and positive forward paths

Static checks:

```bash
flake8 --ignore=E24,W504,E501,C400,C403,C408,C409,C414,C413,C416,C417,C419 \
  nvidia_tao_pytorch/multimodal/clip/loss/masked_siglip_loss.py \
  tests/multimodal_unit_test/clip/test_masked_siglip_loss.py

pydocstyle --ignore=D4,D107,D200,D203,D205,D210,D212,D213,D301,D400,D401 \
  nvidia_tao_pytorch/multimodal/clip/loss/masked_siglip_loss.py

pylint --rcfile .pylintrc \
  nvidia_tao_pytorch/multimodal/clip/loss/masked_siglip_loss.py

python .github/hooks/check_license_header.py \
  nvidia_tao_pytorch/multimodal/clip/loss/masked_siglip_loss.py \
  tests/multimodal_unit_test/clip/test_masked_siglip_loss.py

git diff --check
```

Results:

```text
flake8: passed
pydocstyle: passed
pylint: 10.00/10
license-header check: passed
git diff --check: passed
```

The full CLIP unit-test directory, multi-GPU NCCL execution, and final SFT reproduction have not yet been rerun after the latest BF16 fix.

### Was this patch authored or co-authored using generative AI tooling?

Generated-by: OpenAI Codex (GPT-5)

### Release note

```release-note
Added normalized metadata-compatible positive masking for SigLIP training.
```

### Checklist

- [ ] My commits are signed off per the DCO (`git commit -s`) — see `CONTRIBUTING.md`
- [x] I have read the contributing guidelines
- [x] The code follows the project's style, and lint and format checks pass locally
- [x] I added or updated tests covering this change
- [x] All tests pass locally
- [x] I added or updated documentation (README, docstrings, docs pages, examples)
- [x] I updated the version, changelog, or migration notes if this change requires it
- [x] If this touches components that are optional to install, the imports are guarded so the package still works without them (reviewers: please verify this)
- [x] No secrets, credentials, proprietary data, or customer data are included in this PR
- [x] I understand this contribution is licensed under the repository's license, and that my commit author name and email become permanently public once merged

### Notes for reviewers

Suggested review order:

1. `nvidia_tao_pytorch/multimodal/clip/loss/masked_siglip_loss.py`
2. `nvidia_tao_pytorch/multimodal/clip/model/pl_clip_model.py`
3. `nvidia_tao_pytorch/config/clip/default_config.py`
4. `tests/multimodal_unit_test/clip/test_masked_siglip_loss.py`

Please pay particular attention to distributed loss scaling, FP32 global compatible-pair counting, metadata-evidence gating, and the distinction between `per_pair` and `per_query` normalization.

PA loss, TAL changes, triplet-loss changes, checkpoint changes, and experiment-output changes are outside this PR.

### CODE REVIEW

File: nvidia_tao_pytorch/multimodal/clip/model/pl_clip_model.py [hunk 1]
```diff
@@ -376,13 +391,27 @@ def _build_criterion(self):
...
                     rank=siglip_loss_rank,
                     accessory_aware=(
-                        siglip_mask_mode ==
-                        "attribute_plus_accessory_match_ignore"
+                        siglip_mask_mode in (
+                            "attribute_plus_accessory_match_ignore",
+                            "attribute_plus_accessory_match_positive",
+                        )
+                    ),
+                    compatible_as_positive=compatible_as_positive,
```
Comment (nnagrajrao) [inline nvidia_tao_pytorch/multimodal/clip/model/pl_clip_model.py:412]: Could we block this mode when triplet loss is enabled? The triplet loss still treats every off-diagonal pair as a negative, so the same pair can be trained as both a positive and a negative.

**Review by nnagrajrao — APPROVED** [review]
left one comment that needs to addressed before merging. Approving to unblock

File: nvidia_tao_pytorch/multimodal/clip/model/pl_clip_model.py (hunk 1 above)
Comment (yangchris11) [reply nvidia_tao_pytorch/multimodal/clip/model/pl_clip_model.py:412]: Added a validation guard in f6572e91 that rejects both positive mask modes when train.triplet_loss_weight > 0, with a clear error explaining the conflicting supervision. Added regression tests covering rejection, successful initialization with triplet loss disabled, and unchanged behavior for existing modes. Updated the configuration descriptions in both repos.
